nats/aio/msg.py

```diff
@@ -14,17 +14,27 @@
 
 import datetime
 from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, List, Optional
 from nats.errors import Error, NotJSMessageError, MsgAlreadyAckdError
 
+if TYPE_CHECKING:
+    from nats import NATS
 
+
+@dataclass
```

Comment:
I really like the `dataclass` annotations, do you think there might be some performance impact/benefits from using it for `Msg`?

Reply:
Great question! The dataclasses have `slots` but, unfortunately, only From 3.10, which is not an option just yet for an OSS. So, we can use it only without `slots`. The slotted classes are a little bit faster at attribute lookup and take a bit less memory (minus one dict, basically). You can get a similar benefit by using `typing.NamedTuple`. However, there are a few minor differences:

1. NamedTuple is immutable. You can do the same for dataclasses by setting `frozen=True` but you cannot switch it off for a NamedTuple. It is convenient, sometimes.
2. NamedTuple doesn't allow to use `functools.cached_property`. And this is **very** convenient thing, I can't overstate how nicer and faster it makes the code.
3. NamedTuple behaves like a tuple. It generates `__iter__` and some methods like `count` which may be undesired and confusing for a public API.

Now, back to "the old approach vs dataclasses". My motivation was mostly the convenience of the development since I find the associated performance impact negligible. Dataclasses provide us with:

1. Autogeneration of `__init__`, `__repr__`, and `__eq__`.
2. Again, the ability to use `cached_property`

Now, let's talk about benchmarks. There is  the code I've tested:

```python
from dataclasses import dataclass
class A:
    __slots__ = ['a']
    def __init__(self, a):
        self.a = a
@dataclass
class B:
    a: int

a = A(1)
b = B(1)
```

And there are my results:

```python
%timeit A(1) # 152 ns ± 4.98 ns per loop
%timeit B(1) # 228 ns ± 4.13 ns per loop
%timeit a.a  # 30.8 ns ± 1.1 ns per loop
%timeit b.a  # 33.9 ns ± 1.12 ns per loop
%timeit a.a = 2  # 33.9 ns ± 3.27 ns per loop
%timeit b.a = 2  # 43.7 ns ± 3.87 ns per loop
```

As you can see, there is indeed a difference. Is it negligible? Definitely yes for the lookup operation, and we don't care about assignment one, AFAIK the message is not supposed to be changed. The instantiation, however, gives a noticeable difference, at least in my example. Is it tolerable? I don't know, actually. Depends on the philosophy of the project and where it's heading. For example, I expect it to be even faster if compiled with mypyc. But dataclasses are definitely shorter.

Reply:
Thanks for the explanation!